File: js_implementation/projects/07/VMTranslator-08/CodeWriter.js
// 1. asm里面要写备注like // push local 2
// 2. 避免重复可以写helperFunction
// 3. 结束时写一个infinite loop
const path = require('path');

const base = {
  sp: 256,
  static: 16,
  temp: 5,
  judgeOp: 0,
  funcOp: 0,
  sysOp: 0,
}

const virtualMemorySegments = ['argument', 'local', 'this', 'that'];

let asmResult = [];

// 配置asm指针
const initializer = () => {
  base.sysOp++;
  asmResult.push('// init the stack pointer');
  asmResult.push(`  @${base.sp}`, '  D=A', '  @SP', '  M=D');
  asmResult.push(`  @Sys.INIT${base.sysOp}`, '  D=A');
  pushToStack();
  asmResult.push('  @LCL', '  D=M');
  pushToStack();
  asmResult.push('  @ARG', '  D=M');
  pushToStack();
  asmResult.push('  @THIS', '  D=M');
  pushToStack();
  asmResult.push('  @THAT', '  D=M');
  pushToStack();
  asmResult.push('  @SP', '  D=M');
  asmResult.push('  @5', '  D=D-A', '  @ARG', '  M=D');
  asmResult.push('  @SP', '  D=M', '  @LCL', '  M=D');
  asmResult.push('  @Sys.init', '  0;JMP');
  asmResult.push(`(Sys.INIT${base.sysOp})`)
}

// MAIN
const codeWriter = ({ parsedCommands, isFirstTime, isLastTime, currentFile }) => {

  asmResult = []; // 每一次进来清空数据

  if (isFirstTime) {
    initializer(); //如果第一次进入需要初始化
  }
  parsedCommands.map((commandObj) => {
    if (commandObj?.commandType.indexOf('C_ARITHMETIC') !== -1) {
      writeArithmetic(commandObj);
    } else if (commandObj?.commandType.indexOf('C_PUSH') !== -1 || commandObj?.commandType.indexOf('C_POP') !== -1) {
      writePushPop(commandObj, currentFile);
    } else if (commandObj?.commandType.indexOf('C_BRANCHING') !== -1) {
      writeBranching(commandObj);
    } else if (commandObj?.commandType.indexOf('C_FUNCTION') !== -1) {
      writeFunction(commandObj);
    }
  })
  if (isLastTime) {
    asmResult.push('(END)', '  @END', '  0;JMP');
  }
  const outputContent = asmResult.join('\n');
  return outputContent;
}

const writeArithmetic = (commandObj) => {
  const { arg1 } = commandObj;
  const asmOp = parseArithOp(arg1);
  leaveComment({ arg1 });
  if (arg1 === 'add' || arg1 === 'sub' || arg1 === 'and' || arg1 === 'or') {
    popFromStack();
    asmResult.push('  @R13', '  M=D');
    popFromStack();
    asmResult.push('  @R13', `  D=D${asmOp}M`);
    pushToStack();
  } else if (arg1 === 'neg' || arg1 === 'not') {
    popFromStack();
    asmResult.push(`  M=${asmOp}D`, '  @SP', '  M=M+1');
  } else if (arg1 === 'eq' || arg1 === 'gt' || arg1 === 'lt') {
    base.judgeOp++;
    popFromStack();
    asmResult.push('  @R13', '  M=D');
    popFromStack();
    asmResult.push('  @R13', '  D=D-M');
    asmResult.push(`  @${asmOp}${base.judgeOp}`, `  D;J${asmOp}`, '  @SP', '  A=M', '  M=0', '  @SP', '  M=M+1', `  @CONTINUE${base.judgeOp}`, '  0;JMP');
    asmResult.push(`(${asmOp}${base.judgeOp})`, '  @SP', '  A=M', '  M=-1', '  @SP', '  M=M+1', `(CONTINUE${base.judgeOp})`);
  }
}

const writePushPop = (commandObj, currentFile) => {
  const { commandType, arg1, arg2 } = commandObj;
  if (commandType === 'C_PUSH') {
    leaveComment({ arg1, arg2, op: 'push' });
    if (arg1 === 'constant') {
      encounterPushConstant(arg2);
    } else if (arg1 === 'static') {
      encounterPushStatic(arg2, currentFile);
    } else if (arg1 === 'pointer') {
      encounterPushPointer(arg2);
    } else if (arg1 === 'temp') {
      encounterPushTemp(arg2);
    } else if (virtualMemorySegments.includes(arg1)) {
      const segment = virtualMemorySegments.filter(item => item === arg1)[0];
      encounterPushVirtualMemorySegments(segment, arg2);
    }
  } else if (commandType === 'C_POP') {
    leaveComment({ arg1, arg2, op: 'pop' });
    if (arg1 === 'static') {
      encounterPopStatic(arg2, currentFile);
    } else if (arg1 === 'pointer') {
      encounterPopPointer(arg2);
    } else if (arg1 === 'temp') {
      encounterPopTemp(arg2);
    } else if (virtualMemorySegments.includes(arg1)) {
      const segment = virtualMemorySegments.filter(item => item === arg1)[0];
      encounterPopVirtualMemorySegments(segment, arg2);
    }
  }
}

const writeBranching = (commandObj) => {
  const { commandType, arg1 } = commandObj;
  if (commandType === 'C_BRANCHING_LABEL') {
    leaveComment({ arg1, op: 'label' });
    asmResult.push(`(${arg1})`);
  } else if (commandType === 'C_BRANCHING_IFGOTO') {
    leaveComment({ arg1, op: 'if-goto' });
    popFromStack();
    asmResult.push(`  @${arg1}`, '  D;JNE');
  } else if (commandType === 'C_BRANCHING_GOTO') {
    leaveComment({ arg1, op: 'goto' });
    asmResult.push(`  @${arg1}`, '  0;JMP');
  }
}

const writeFunction = (commandObj) => {
  const { commandType, arg1, arg2 } = commandObj;
  if (commandType === 'C_FUNCTION_DECLARE') {
    leaveComment({ arg1, arg2, op: 'declare' });
    asmResult.push(`(${arg1})`);
    for (let i = 0; i < arg2; i++) {
      asmResult.push(`  @${i}`, '  D=A', '  @LCL', '  A=M+D', '  M=0');
    }
  } else if (commandType === 'C_FUNCTION_RETURN') {
    leaveComment({ arg1, op: 'function' });
    asmResult.push('  @LCL', '  D=M', '  @R13', '  M=D'); //用R13(frame)存储LCL的值
    asmResult.push('  @5', '  D=A', '  @R13', '  A=M-D', '  D=M', '  @R14', '  M=D'); // R14(retAddr)存储返回的跳转地址
    popFromStack();
    asmResult.push('  @ARG', '  A=M', '  M=D'); // 让ARG[0]存储func返回值
    asmResult.push('  @ARG', '  D=M+1', '  @SP', '  M=D'); // 让SP指向返回值位置的下一个地址(ARG[1])
    asmResult.push('  @1', '  D=A', '  @R13', '  A=M-D', '  D=M', '  @THAT', '  M=D'); // reinstate THAT指针
    asmResult.push('  @2', '  D=A', '  @R13', '  A=M-D', '  D=M', '  @THIS', '  M=D'); // reinstate THIS指针
    asmResult.push('  @3', '  D=A', '  @R13', '  A=M-D', '  D=M', '  @ARG', '  M=D'); // reinstate ARG指针
    asmResult.push('  @4', '  D=A', '  @R13', '  A=M-D', '  D=M', '  @LCL', '  M=D'); // reinstate LCL指针
    asmResult.push('  @R14', '  A=M', '  0;JMP'); // 跳转返回地址
  } else if (commandType === 'C_FUNCTION_CALL') {
    leaveComment({ arg1, arg2, op: 'call' });
    base.funcOp++;
    // push frame into stack
    asmResult.push(`  @${arg1}RET${base.funcOp}`, '  D=A');
    pushToStack();
    asmResult.push('  @LCL', '  D=M');
    pushToStack();
    asmResult.push('  @ARG', '  D=M');
    pushToStack();
    asmResult.push('  @THIS', '  D=M');
    pushToStack();
    asmResult.push('  @THAT', '  D=M');
    pushToStack();
    // reposite the ARG
    asmResult.push('  @SP', '  D=M');
    asmResult.push(`  @${arg2 + 5}`, '  D=D-A', '  @ARG', '  M=D');
    // LCL points to SP
    asmResult.push('  @SP', '  D=M', '  @LCL', '  M=D'); 
    // goto the callee
    asmResult.push(`  @${arg1}`, '  0;JMP');
    asmResult.push(`(${arg1}RET${base.funcOp})`);
  }
}

// HELPER
// 1. push or pop
const pushToStack = () => {
  asmResult.push('  @SP', '  A=M', '  M=D', '  @SP', '  M=M+1');
};

const popFromStack = () => {
  asmResult.push('  @SP', '  AM=M-1', '  D=M');
}

const encounterPushConstant = (value) => {
  asmResult.push(`  @${value}`, '  D=A');
  pushToStack();
}

const encounterPushStatic = (value, currentFile) => {
  asmResult.push(`  @${currentFile}.${value}`, '  D=M');
  pushToStack();
}

const encounterPushTemp = (value) => {
  asmResult.push(`  @${base.temp + value}`, '  D=M');
  pushToStack();
}

const encounterPushPointer = (value) => {
  if (value === 0) {
    asmResult.push('  @THIS', '  D=M');
    pushToStack();
  } else if (value === 1) {
    asmResult.push('  @THAT', '  D=M');
    pushToStack();
  }
}

const encounterPushVirtualMemorySegments = (segment, value) => {
  const asmSegment = parseSegment(segment);
  asmResult.push(`  @${value}`, '  D=A', `  @${asmSegment}`, '  A=M+D', '  D=M');
  pushToStack();
}

const encounterPopStatic = (value, currentFile) => {
  popFromStack();
  asmResult.push(`  @${currentFile}.${value}`, '  M=D');
}

const encounterPopTemp = (value) => {
  popFromStack();
  asmResult.push(`  @${base.temp + value}`, '  M=D');
}

const encounterPopPointer = (value) => {
  if (value === 0) {
    popFromStack();
    asmResult.push('  @THIS', '  M=D');
  } else if (value === 1) {
    popFromStack();
    asmResult.push('  @THAT', '  M=D');
  }
}

const encounterPopVirtualMemorySegments = (segment, value) => {
  const asmSegment = parseSegment(segment);
  asmResult.push(`  @${value}`, '  D=A', `  @${asmSegment}`, '  D=M+D', '  @R13', '  M=D');
  popFromStack();
  asmResult.push('  @R13', '  A=M', '  M=D');
}

const parseSegment = (segment) => {
  switch (segment) {
    case 'argument':
      return 'ARG';
    case 'local':
      return 'LCL';
    case 'this':
      return 'THIS';
    case 'that':
      return 'THAT';
  }
};

// 2. arithmetic

const parseArithOp = (op) => {
  switch (op) {
    case 'add':
      return '+';
    case 'sub':
      return '-';
    case 'and':
      return '&';
    case 'or':
      return '|';
    case 'neg':
      return '-';
    case 'not':
      return '!';
    case 'eq':
      return 'EQ';
    case 'lt':
      return 'LT';
    case 'gt':
      return 'GT';
  }
}

// 3. comment
const leaveComment = (params) => {
  const { op, arg1, arg2 } = params;
  if (arg2 !== undefined && op !== undefined) {
    asmResult.push(`// ${op} ${arg1} ${arg2}`);
  } else if (arg1 !== undefined && op !== undefined) {
    asmResult.push(`// ${op} ${arg1}`);
  } else {
    asmResult.push(`// ${arg1}`);
  }
}

module.exports = codeWriter;

File: src/project7/src/code_writer.c
#include "code_writer.h"
#include <string.h>

int CONDITION_STATEMENT_COUNTER = 0;

void write_comment(FILE *file, Statement *s)
{
    fprintf(file, "// ");

    switch (s->cmd) {
    case C_ARITHMETIC:
    case C_RETURN:
        fprintf(file, "%s\n", s->arg1);
        break;
    case C_PUSH:
        fprintf(file, "push %s %d\n", s->arg1, s->arg2);
        break;
    case C_POP:
        fprintf(file, "pop %s %d\n", s->arg1, s->arg2);
        break;
    case C_FUNCTION:
        fprintf(file, "function %s %d\n", s->arg1, s->arg2);
        break;
    case C_CALL:
        fprintf(file, "call %s %d\n", s->arg1, s->arg2);
        break;
    default:
        break;
    }
}

void pop_stack_value_to_D_register(FILE *file)
{
    fprintf(file, "\t@SP\n");
    fprintf(file, "\tAM=M-1\n");
    fprintf(file, "\tD=M\n");
}

void copy_D_register_value_to(FILE *file, const char* dst)
{
    fprintf(file, "\t@%s\n", dst);
    fprintf(file, "\tM=D\n");
}

void add_D_register_value_with(FILE *file, const char* src)
{
    fprintf(file, "\t@%s\n", src);
    fprintf(file, "\tD=D+M\n");
}

void sub_D_register_value_with(FILE *file, const char* src)
{
    fprintf(file, "\t@%s\n", src);
    fprintf(file, "\tD=D-M\n");
}

void and_D_register_value_with(FILE *file, const char* src)
{
    fprintf(file, "\t@%s\n", src);
    fprintf(file, "\tD=D&M\n");
}

void or_D_register_value_with(FILE *file, const char* src)
{
    fprintf(file, "\t@%s\n", src);
    fprintf(file, "\tD=D|M\n");
}

void push_D_register_value_to_stack(FILE *file)
{
    fprintf(file, "\t@SP\n");
    fprintf(file, "\tM=D\n");

    fprintf(file, "\t@SP\n");
    fprintf(file, "\tM=M+1\n");
}


void copy_segment_value_to_D_register(FILE *file, Statement *s, const char *vm_file_root)
{
    if (strcmp(s->arg1, "constant") == 0) {
        fprintf(file, "\t@%d\n", s->arg2);
        fprintf(file, "\tD=A\n");
        return;
    }

    if (strcmp(s->arg1, "static") == 0) {
        fprintf(file, "\t@%s.%d\n", vm_file_root, s->arg2);
        fprintf(file, "\tD=M\n");
        return;
    }

    if (strcmp(s->arg1, "pointer") == 0) {
        fprintf(file, "\t@%s\n", (s->arg2 == 0) ? "THIS" : "THAT");
        fprintf(file, "\tD=M\n");
        return;
    }

    if (strcmp(s->arg1, "temp") == 0) {
        fprintf(file, "\t@R%d\n", (5 + s->arg2));
        fprintf(file, "\tD=M\n");
        return;
    }

    if (strcmp(s->arg1, "local") == 0) {
        fprintf(file, "\t@LCL\n");
    } else if (strcmp(s->arg1, "argument") == 0) {
        fprintf(file, "\t@ARG\n");
    } else if (strcmp(s->arg1, "this") == 0) {
        fprintf(file, "\t@THIS\n");
    } else if (strcmp(s->arg1, "that") == 0) {
        fprintf(file, "\t@THAT\n");
    }

    fprintf(file, "\tD=M\n");
    fprintf(file, "\t@%d\n", s->arg2);
    fprintf(file, "\tA=D+A\n");
    fprintf(file, "\tD=M\n");
}

void write_arithmetic(FILE *file, Statement *s)
{
    write_comment(file, s);
    pop_stack_value_to_D_register(file);

    if (strcmp(s->arg1, "add") == 0) {
        copy_D_register_value_to(file, TEMP_REGISTER);
        pop_stack_value_to_D_register(file);
        add_D_register_value_with(file, TEMP_REGISTER);
        push_D_register_value_to_stack(file);
        goto ret;
    }

    if (strcmp(s->arg1, "sub") == 0) {
        copy_D_register_value_to(file, TEMP_REGISTER);
        pop_stack_value_to_D_register(file);
        sub_D_register_value_with(file, TEMP_REGISTER);
        push_D_register_value_to_stack(file);
        goto ret;
    }

    if (strcmp(s->arg1, "neg") == 0) {
        fprintf(file, "\tD=-D\n");
        push_D_register_value_to_stack(file);
        goto ret;
    }

    if (strcmp(s->arg1, "eq") == 0) {
        copy_D_register_value_to(file, TEMP_REGISTER);
        pop_stack_value_to_D_register(file);
        sub_D_register_value_with(file, TEMP_REGISTER);
        fprintf(file, "\t@EQ%d\n", ++CONDITION_STATEMENT_COUNTER);
        fprintf(file, "\tD;JEQ\n");
        fprintf(file, "\tD=0\n");
        push_D_register_value_to_stack(file);
        fprintf(file, "\t@NEXT%d\n", CONDITION_STATEMENT_COUNTER);
        fprintf(file, "\t0;JMP\n");
        fprintf(file, "(EQ%d)\n", CONDITION_STATEMENT_COUNTER);
        fprintf(file, "\tD=-1\n");
        push_D_register_value_to_stack(file);
        fprintf(file, "(NEXT%d)\n", CONDITION_STATEMENT_COUNTER);
    }

    if (strcmp(s->arg1, "gt") == 0) {
        copy_D_register_value_to(file, TEMP_REGISTER);
        pop_stack_value_to_D_register(file);
        sub_D_register_value_with(file, TEMP_REGISTER);
        fprintf(file, "\t@GT%d\n", ++CONDITION_STATEMENT_COUNTER);
        fprintf(file, "\tD;JGT\n");
        fprintf(file, "\tD=0\n");
        push_D_register_value_to_stack(file);
        fprintf(file, "\t@NEXT%d\n", CONDITION_STATEMENT_COUNTER);
        fprintf(file, "\t0;JMP\n");
        fprintf(file, "(GT%d)\n", CONDITION_STATEMENT_COUNTER);
        fprintf(file, "\tD=-1\n");
        push_D_register_value_to_stack(file);
        fprintf(file, "(NEXT%d)\n", CONDITION_STATEMENT_COUNTER);
    }

    if (strcmp(s->arg1, "lt") == 0) {
        copy_D_register_value_to(file, TEMP_REGISTER);
        pop_stack_value_to_D_register(file);
        sub_D_register_value_with(file, TEMP_REGISTER);
        fprintf(file, "\t@LT%d\n", ++CONDITION_STATEMENT_COUNTER);
        fprintf(file, "\tD;JLT\n");
        fprintf(file, "\tD=0\n");
        push_D_register_value_to_stack(file);
        fprintf(file, "\t@NEXT%d\n", CONDITION_STATEMENT_COUNTER);
        fprintf(file, "\t0;JMP\n");
        fprintf(file, "(LT%d)\n", CONDITION_STATEMENT_COUNTER);
        fprintf(file, "\tD=-1\n");
        push_D_register_value_to_stack(file);
        fprintf(file, "(NEXT%d)\n", CONDITION_STATEMENT_COUNTER);
    }

    if (strcmp(s->arg1, "and") == 0) {
        copy_D_register_value_to(file, TEMP_REGISTER);
        pop_stack_value_to_D_register(file);
        and_D_register_value_with(file, TEMP_REGISTER);
        push_D_register_value_to_stack(file);
    }

    if (strcmp(s->arg1, "or") == 0) {
        copy_D_register_value_to(file, TEMP_REGISTER);
        pop_stack_value_to_D_register(file);
        or_D_register_value_with(file, TEMP_REGISTER);
        push_D_register_value_to_stack(file);
    }

    if (strcmp(s->arg1, "not") == 0) {
        fprintf(file, "\tD=!D\n");
        push_D_register_value_to_stack(file);
        goto ret;
    }

ret:
    return;
}

void code_writer(FILE *file, Statement *s, const char *vm_file_root)
{
    switch (s->cmd) {
    case C_ARITHMETIC:
        write_arithmetic(file, s);
        break;
    case C_PUSH:
        write_comment(file, s);
        copy_segment_value_to_D_register(file, s, vm_file_root);
        push_D_register_value_to_stack(file);
        break;
    default:
        break;
    }
}
